File: src/0019_remove_nth_node_from_end_of_list/remove_nth_node_from_end_of_list.cc
// author: yqq
// date: 2021-05-24 19:33:09
// descriptions: https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list/
#include <iostream>
#include <vector>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode* retHead = head;
        ListNode* p1 = head,  *p2 = head; 
        int i = 0;
        for(; i < n && p1->next != nullptr; i++)
        {
            p1 = p1->next;
        }
        p1 = p1->next;

        // 删除head本身
        if(p1 == nullptr && i == n-1) 
        {
            return head->next;
        }

        for(;p1 != nullptr;)
        {
            p1 = p1->next;
            p2 = p2->next;
        }

        p2->next = p2->next->next;
        return retHead;
    }
};

int main()
{
    // [1,2,3,4,5],
    // [1]
    // [1,2]
    // ListNode node1(1, nullptr);

    // ListNode node2(2, nullptr);
    // ListNode node1(1, &node2);

    // [1,2,3]
    // ListNode node3(3, nullptr );
    // ListNode node2(2, &node3);
    // ListNode node1(1, &node2);
    
    ListNode node5(5, nullptr );
    ListNode node4(4, &node5);
    ListNode node3(3, &node4 );
    ListNode node2(2, &node3);
    ListNode node1(1, &node2);

    ListNode *header = &node1;
    Solution sol;
    // ListNode* head = sol.removeNthFromEnd(header, 2);
    ListNode* head = sol.removeNthFromEnd(header, 2);

    for(ListNode* p = head; p != nullptr; p = p->next)
    {
        cout << p->val << ",";
    }
    cout << endl;

    // cout << "hello world" << endl;
    return 0;
}

File: src/0019_remove_nth_node_from_end_of_list/remove_nth_node_from_end_of_list.rs
// author: yqq
// date: 2021-05-24 19:33:09
// descriptions: https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

struct Solution {}

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {

        let retHead = &head;
        let mut p1 = &head; 
        let mut p2 = &head ;

        let mut i = 0;
        while i < n && !p1.is_none() {
            p1 = &p1.as_ref().unwrap().next;
            i += 1;
        }
        p1 = &p1.as_ref().unwrap().next;

        if(p1.is_none() && i == n - 1) {
            return head.unwrap().next;
        }

        while !p1.is_none() {
            p1 = &p1.as_ref().unwrap().next;
            p2 = &p2.as_ref().unwrap().next;
        }

        // p2.as_ref().unwrap().next.as_mut() = p2.as_ref().unwrap().next.unwrap().next;
        return *retHead.as_ref();
    }
}


fn main(){

    let mut node2: ListNode = ListNode::new(2);
    let mut node1: ListNode = ListNode::new(1);
    node1.next = Some(Box::new(node2));


    println!("hello world")
}

